twister: Add support for Cpputest

vChavezB opened the pull request:
Similar to gTest, CppuTest is a CPP framework for unit tests. This commit adds support to detect Cpputest console output when verbose mode is enabled (-v).

I am open for any feedback. My main use case was some unit tests that were written with Cpputest that I needed to port to Zephyr and integrate with twister.

MaureenHelm commented:
@vChavezB please take a look at the failed checks

vChavezB commented:
Thanks for the feedback, I must have missed it.

hakehuang commented:
would be possible to a test application for this?

PerMac commented on doc/develop/test/twister.rst:
```diff
@@ -459,15 +459,17 @@ harness: <string>
+    output and matching certain phrases. ``ztest``, ``gtest`` and ``cpputest`` harnesses look
+    for pass/fail/etc. frames defined in those frameworks. Use ``gtest`` or ``cpputest``
+    harness if you've already got tests written with either of these framework and do
+    not wish to update them to zTest. If using ``cpputest`` be sure to pass the
+    verbose argument ``-v`` to `CommandLineTestRunner::RunAllTests`.
```
Why the need for "-v"? I don't see it in the code. Also, it shouldn't be the case. Level of output verbosity mustn't affect a test result.

vChavezB replied:
The verbose argument must be applied to your  CPPUtest application. Correct me if I am wrong but if you do not add the verbose option, you do not get information about which test cases have passed. Instead you will only get a dot for each test that has passed (e.g. `....`) and at the end something such as ` OUTPUT: OK (2 tests, 2 ran, 94 checks, 0 ignored, 0 filtered out, 2 ms)`.

vChavezB commented:
> would be possible to a test application for this?

I made a sample here

https://github.com/vChavezB/zephyr_cpputest_sample

```
git clone https://github.com/vChavezB/zephyr_cpputest_sample
wes init -l zephyr_cpputest_sample
west update -o=--depth=1 -n
./zephyr/scripts/twister -vvv -T zephyr_cpputest_sample
```

hakehuang commented:
> > would be possible to a test application for this?
> 
> I made a sample here
> 
> https://github.com/vChavezB/zephyr_cpputest_sample
> 
> ```
> git clone https://github.com/vChavezB/zephyr_cpputest_sample
> wes init -l zephyr_cpputest_sample
> west update -o=--depth=1 -n
> ./zephyr/scripts/twister -vvv -T zephyr_cpputest_sample
> ```

I would support you to move this test into tests/cpputest

vChavezB commented:
I have rebased with last changes to twister script. Will add as suggested by @hakehuang  the test sample to `tests/cpputest`.

vChavezB commented:
I would like the stale label removed. Still waiting for more reviews.

mmahadevan108 commented:
@PerMac @nashif , can you please take a look at this PR.

nashif commented:
looks like testcases are not being captured correctly...

```
INFO    - Added initial list of jobs to queue
ERROR   - A None status detected in instance native_sim/native/64/tests/cpputest/base/base.my_test_group, test case base.my_test_group.
INFO    - 1/1 native_sim/native/64      tests/cpputest/base/base.my_test_group             PASSED (native 0.009s)
```

vChavezB commented:
> looks like testcases are not being captured correctly...
> 
> ```
> INFO    - Added initial list of jobs to queue
> ERROR   - A None status detected in instance native_sim/native/64/tests/cpputest/base/base.my_test_group, test case base.my_test_group.
> INFO    - 1/1 native_sim/native/64      tests/cpputest/base/base.my_test_group             PASSED (native 0.009s)
> ```

Could you tell me how to reproduce the error ?

I ran 

./zephyr/scripts/twister -T tests/cpputest/

and did not get the error, should I rebase ?

```
ZEPHYR_BASE unset, using "/home/ubu/z_cpputest"
Renaming output directory to /home/ubu/z_cpputest/twister-out.1
INFO    - Using Ninja..
INFO    - Zephyr version: 4c4456c76bea
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting default platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/ubu/z_cpputest/twister-out/testplan.json
INFO    - JOBS: 8
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    1/   1  100%  skipped:    0, failed:    0, error:    0
INFO    - 1 test scenarios (1 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 errored, 0 skipped with 0 warnings in 29.18 seconds
INFO    - In total 2 test cases were executed, 0 skipped on 1 out of total 772 platforms (0.13%)
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/ubu/z_cpputest/twister-out/twister.json
INFO    - Writing xunit report /home/ubu/z_cpputest/twister-out/twister.xml...
INFO    - Writing xunit report /home/ubu/z_cpputest/twister-out/twister_report.xml...
INFO    - Run completed
```

nashif commented:
yes, you should rebase.

vChavezB commented:
This are the test cases that are run by twister

```
INFO    - 1/1 native_sim/native/64      tests/cpputest/base/base.my_test_group             PASSED (native 0.005s)
INFO    -                                    base.my_test_group                                                          NONE
INFO    -                                    base.my_test_group.my_test_group.test_1                                     PASSED
```

The only test case that I defined in the sample code is `test_1 , see [here](https://github.com/vChavezB/zephyr/blob/13d0b8c58b29cc58f67ba59fb3252e92ca04bf46/tests/cpputest/base/src/test_suite.cpp#L20-L22)

```cpp
TEST(my_test_group, test_1) {

}
```

However, it seems that twister uses the name `base.my_test_group`(see [here ](https://github.com/vChavezB/zephyr/blob/13d0b8c58b29cc58f67ba59fb3252e92ca04bf46/tests/cpputest/base/testcase.yaml#L6) ) as a test case.

```yaml
tests:
  base.my_test_group:
    platform_allow:
      - native_sim/native/64
    harness: "cpputest"
```

This is implemented [here](https://github.com/vChavezB/zephyr/blob/c5fa9af23569fc7fc081c4896e968a22806b0d58/scripts/pylib/twister/twisterlib/testinstance.py#L120).
```python
    # Fix an issue with copying objects from testsuite, need better solution.
    def init_cases(self):
        for c in self.testsuite.testcases:
            self.add_testcase(c.name, freeform=c.freeform)
```

This has confused me...

Should the test names I defined in the yaml be considered as tests suites/groups or as individual test cases?

vChavezB commented:
I have opened an issue with the error that you mentioned. This is related to how test cases are parsed for other frameworks that are not ztest.

https://github.com/zephyrproject-rtos/zephyr/issues/81176

vChavezB commented:
I have made a temporal solution to allow that my CPPUTest harness class deletes the fake testcase at start, but I would opt for fixing the issue in #81176 .

vChavezB commented:
Rebased and fixed according to https://github.com/zephyrproject-rtos/zephyr/issues/81176#issuecomment-2604478074

nashif commented on tests/cpputest/base/testcase.yml:
```diff
@@ -0,0 +1,9 @@
+common:
```
rename this file to testcase.yaml please

vChavezB replied:
thanks for the feedback. Have changed now the extension to yaml.

yperess commented on scripts/tests/twister/test_harness.py:
```diff
@@ -1209,6 +1221,124 @@ def test_gtest_repeated_run(gtest):
+    assert cpputest.instance.get_case_by_name("id.suite_name.test_name").status == TwisterStatus.FAIL
+
+
+def test_cpputest_test_repeated(cpputest):
```
What if I want to call `RunAllTests` , multiple times?

yperess replied:
Oops, I probably wrote it. Give me a bit, I'll take a look and get back to you later today.

yperess replied:
Marking as resolved, sorry it took so long.

hakehuang commented on tests/cpputest/base/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+include(FetchContent)
+FetchContent_Declare(
+    CppUTest
+    GIT_REPOSITORY https://github.com/cpputest/cpputest.git
```
is it possible to move cpputest project into zephyr umbrella? further, ztest has many configs which would be good to aligned with cpputest as well.

vChavezB replied:
By moving, do you mean declaring a west manifest under `tests/cpputest/base/west.yml` with `cpputest `or something such as creating a repository under the github url `zephyproject-rtos/cpputest` ?

about the configs you mean creating an interoperability layer to make the KConfigs from ZTest compatible when using cpputest such as

- `https://docs.zephyrproject.org/latest/develop/test/ztest.html#repeating-tests`

hakehuang replied:
> By moving, do you mean declaring a west manifest under `tests/cpputest/base/west.yml` with `cpputest `or something such as creating a repository under the github url `zephyproject-rtos/cpputest` ?

yes

hakehuang replied:
> about the configs you mean creating an interoperability layer to make the KConfigs from ZTest compatible when using cpputest such as
> 
> * https://docs.zephyrproject.org/latest/develop/test/ztest.html#repeating-tests

check this file,
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/testsuite/ztest/Kconfig

vChavezB replied:
> > By moving, do you mean declaring a west manifest under `tests/cpputest/base/west.yml` with `cpputest `or something such as creating a repository under the github url `zephyproject-rtos/cpputest` ?
> 
> yes

Do you mean yes...declaring a west manifest or yes creating a repository?  If you mean declaring a west manifest, do you mean adding cpputest to the root of the zephyr project repo, such as is the case for `bsim`, see https://github.com/zephyrproject-rtos/zephyr/blob/961593d1529002b33cb568499090c111584d7c42/west.yml#L36

> check this file,
> https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/testsuite/ztest/Kconfig

Are you proposing to add a repository under `/subsys/testsuite/cpputest` and replicate the Kconfigs from ztest for cpputest? 

sorry for the specific questions but its not that clear for me as I havent used ztest.

hakehuang replied:
> Do you mean yes...declaring a west manifest or yes creating a repository? If you mean declaring a west manifest, do you mean adding cpputest to the root of the zephyr project repo, such as is the case for `bsim`, see

I would suggest to create a repo and add to west manifest. so that the cpputest is a module in zephyr project.

hakehuang replied:
> Are you proposing to add a repository under `/subsys/testsuite/cpputest` and replicate the Kconfigs from ztest for cpputest?
> 
> sorry for the specific questions but its not that clear for me as I havent used ztest.

I am not sure, all the ztest configs applys to cpputest, but would that be possible that you can reuse all the configs? for example, we do not what to change the test stack setting twice.

yperess replied:
I don't know how applicable that is, are you expecting a single binary to contain both ztest and cpputest?

hakehuang replied:
no actually. I will raise this problem in testing working group this week. Maybe we can enhance in following PR, but would this possible that you create a similar Kconfig for cpputest first.

hakehuang replied:
per TWG meeting, we agreed that Kconfig is no need to aligned with ztest.

vChavezB replied:
thanks for the info. So what would be required to continue with the PR. 
Kind regards

hakehuang replied:
@vChavezB  can you contact @nashif on whether cpputest can move into zephyr project.

vChavezB replied:
Is it enough to tag him here? @nashif , as per @hakehuang , there is a question regarding adding cpputest to the zephyr project.

From my perspective unless, someone will mantain cpputest for the zephyr project it is a good idea to add it. If not I would suggest just adding a west manifest to `tests/cpputest/base/west.yml` if using the CMake `FetchContent_Declare` is not preferred by the maintainers.

vChavezB committed:
twister: Add support for Cpputest

Similar to gTest, CppuTest is a CPP framework for unit tests.
This commit adds support based on the console output of
a cpputest test suite application.

Signed-off-by: Victor Chavez <vchavezb@protonmail.com>

vChavezB commented:
rebased as there was a conflict in the docs with the addition of harness "power". Will leave this PR for any further discussion on what to with cpputest as mentioned here https://github.com/zephyrproject-rtos/zephyr/pull/74372#discussion_r2040637745

epc-ake commented:
Are there still plans to merge this?
I would appreciate cpputest support in zephyr since a lot of our existing code bases test using cpputest.
